### PR TITLE
Azure blob storage service

### PR DESCRIPTION
**Pull request by aviramha**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -15,6 +15,8 @@ yaspin = "^1.0.0"
 redis = {version="*", optional=true}
 pika = {version="*", optional=true}
 kafka-python = {version="*", optional=true}
+azure-storage-blob = { version = "^12.3.1", optional = true }
+cffi = {version = "^1.14.0", optional = true}
```
Comment (bentheiii) [inline pyproject.toml:19]: should be * to support compatibility

File: tests/extras/test_azure_storage.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+@mark.parametrize('spinner', [True, False])
+def test_make_azure_storage(docker_client, spinner):
+    with BlobStorageService.run(docker_client, spinner=spinner):
+        pass
+
+def test_upload_download_files(docker_client):
+    with BlobStorageService.run(docker_client) as service:
+        service.upload_files("test", {"file_1": b"data", "file_2": b"data2"})
+        assert service.download_file("test", "file_1") == b"data"
+        assert service.download_file("test", "file_2") == b"data2"
```
Comment (bentheiii) [inline tests/extras/test_azure_storage.py:16]: add tests to connect to from sibling containers, both from network and from docker host (see rabbit)

**Review by bentheiii — CHANGES_REQUESTED** [review]

**Comment by bentheiii** [comment]
to fix the tests, add `-E azure` to the github actions

File: tests/extras/test_azure_storage.py (hunk 2 above)
Comment (aviramha) [reply tests/extras/test_azure_storage.py:16]: what would be the difference between this container and other containers in those regards?

File: pyproject.toml (hunk 1 above)
Comment (bharel) [reply pyproject.toml:19]: Dunno, internal stuff might not work in case of breakage.

File: tests/extras/test_azure_storage.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            url = f"http://{aliases[0]}:{BLOB_STORAGE_DEFAULT_PORT}"
+            container = create_and_pull(
+                docker_client,
+                "byrnedo/alpine-curl", f'-vvv -I "{url}" --http0.9',
+                detach=True
+            )
+            with connect(network, container):
+                container.start()
+                return_status = container.wait()
+                assert return_status["StatusCode"] == 0
```
Comment (bentheiii) [inline tests/extras/test_azure_storage.py:41]: add sibling test through docker-host

File: yellowbox/extras/azure_storage.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+
+    def stop(self, signal: str = 'SIGKILL'):
+        """
+        We override to change the signal.
+        """
+        super().stop(signal)
+
+    def client_port(self):
+        return get_ports(self.container)[BLOB_STORAGE_DEFAULT_PORT]
```
Comment (bentheiii) [inline yellowbox/extras/azure_storage.py:38]: newline at end of file

File: yellowbox/extras/azure_storage.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+class BlobStorageService(SingleContainerService, RunnableWithContext):
+    """
+    Starts Azurite, Azure's storage emulator.
+    Provides helper functions for preparing the instance for testing.
+    TODO: Make account name and key configurable.
+    """
+
+    def __init__(self, docker_client,
+                 image="mcr.microsoft.com/azure-storage/azurite:latest",
+                 *, remove=True):
```
Comment (bentheiii) [inline yellowbox/extras/azure_storage.py:25]: my god, it's full of newlines

Comment (bentheiii) [inline yellowbox/extras/azure_storage.py:25]: use **kwargs instead

File: yellowbox/extras/azure_storage.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        container = create_and_pull(
+            docker_client, image, "azurite-blob --blobHost 0.0.0.0", publish_all_ports=True)
+        super().__init__(container, remove=remove)
+
+
+    def stop(self, signal: str = 'SIGKILL'):
```
Comment (bentheiii) [inline yellowbox/extras/azure_storage.py:31]: signal is not str (could also be an int)

**Review by bentheiii — CHANGES_REQUESTED** [review]

**Comment by bentheiii** [comment]
should be merged after #20 since we rename RunnableWithContext

File: yellowbox/extras/azure_storage.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+class ResourceNotReady(Exception):
+    pass
+
+
+def storage_is_ready(container: Container) -> None:
+    """
+    Checks if storage is ready, raises exception if not
+    """
+    if b"Azurite Blob service successfully listens on" not in container.logs():
+        raise ResourceNotReady
```
Comment (bentheiii) [inline yellowbox/extras/azure_storage.py:32]: raise RuntimeError

Comment (bentheiii) [inline yellowbox/extras/azure_storage.py:32]: should be inner function in start

**Review by bentheiii — CHANGES_REQUESTED** [review]

**Review by bentheiii — DISMISSED** [review]

**Review by bentheiii — APPROVED** [review]

**Review by bentheiii — APPROVED** [review]

**Pull request merged**